src/bridging/providers/across/AcrossBridgeProvider.ts

```diff
@@ -0,0 +1,271 @@
+
+import {
+  BridgeDeposit,
+  BridgeHook,
+  BridgeProvider,
+  BridgeProviderInfo,
+  BridgeQuoteResult,
+  BridgeStatusResult,
+  GetBuyTokensParams,
+  QuoteBridgeRequest,
+} from '../../types'
+
+import { RAW_PROVIDERS_FILES_PATH } from '../../const'
+
+import { ChainId, ChainInfo, SupportedChainId, TargetChainId } from '../../../chains'
+
+import { ACROSS_TOKEN_MAPPING } from './const/tokens'
+import { EvmCall, TokenInfo } from '../../../common'
+
+import { mainnet } from '../../../chains/details/mainnet'
+import { polygon } from '../../../chains/details/polygon'
+import { arbitrumOne } from 'src/chains/details/arbitrum'
+import { base } from '../../../chains/details/base'
+import { optimism } from '../../../chains/details/optimism'
+import { AcrossApi, AcrossApiOptions, SuggestedFeesResponse } from './AcrossApi'
+import { getChainConfigs, getTokenAddress, getTokenSymbol, toBridgeQuoteResult } from './util'
+import { ACROSS_MATH_CONTRACT_ADDRESSES, ACROSS_SPOOK_CONTRACT_ADDRESSES } from './const/contracts'
+import { CowShedSdk, CowShedSdkOptions } from '../../../cow-shed'
+import { CommandFlags, createWeirollDelegateCall } from '../../../weiroll'
+import { ACROSS_MATH_ABI, ACROSS_SPOKE_POOL_ABI } from './abi'
+
+const HOOK_DAPP_ID = 'cow-sdk://bridging/providers/across'
+const ERC20_BALANCE_OF_ABI = ['function balanceOf(address account) external view returns (uint256)'] as const
+
+const ERC20_APPROVE_OF_ABI = ['function approve(address spender, uint256 amount) external returns (bool)'] as const
+export const ACROSS_SUPPORTED_NETWORKS = [mainnet, polygon, arbitrumOne, base, optimism]
+
+export interface AcrossBridgeProviderOptions {
+  /**
+   * Token info provider
+   * @param chainId - The chain ID
+   * @param addresses - The addresses of the tokens to get the info for
+   * @returns The token infos
+   */
+  getTokenInfos: (chainId: ChainId, addresses: string[]) => Promise<TokenInfo[]>
+
+  // API options
+  apiOptions?: AcrossApiOptions
+
+  // Cow-shed options
+  cowShedOptions?: CowShedSdkOptions
+}
+
+export interface AcrossQuoteResult extends BridgeQuoteResult {
+  suggestedFees: SuggestedFeesResponse
+}
+
+export class AcrossBridgeProvider implements BridgeProvider<AcrossQuoteResult> {
+  protected api: AcrossApi
+  protected cowShedSdk: CowShedSdk
+
+  constructor(private options: AcrossBridgeProviderOptions) {
+    this.api = new AcrossApi(options.apiOptions)
+    this.cowShedSdk = new CowShedSdk(options.cowShedOptions)
+  }
+
+  info: BridgeProviderInfo = {
+    name: 'Across',
+    logoUrl: `${RAW_PROVIDERS_FILES_PATH}/across/across-logo.png`,
+  }
+
+  async getNetworks(): Promise<ChainInfo[]> {
+    return ACROSS_SUPPORTED_NETWORKS
+  }
+
+  async getBuyTokens(param: GetBuyTokensParams): Promise<TokenInfo[]> {
+    const { targetChainId } = param
+
+    const chainConfig = ACROSS_TOKEN_MAPPING[targetChainId as TargetChainId]
+    if (!chainConfig) {
+      return []
```

Comment:
Nitpick.
With the empty array we cannot differentiate if a network is unsupported or it's supported, but there are zero tokens for given GetBuyTokensParams.
Maybe we should return null instead.

Reply:
I can, and is not a bad idea, but opted for simpler interface since we already have a way to see the supported networks `provider.getNetworks()`

Reply:
Let me know if you still want me to add it, I just feel the provider interface gives the same info, so I don't want to make the clients handle the `null` if they already only ask for the tokens for one of those networks